File: plugin/src/main/groovy/ru/cleverpumpkin/badge/ImageWriter.java
package ru.cleverpumpkin.badge;


import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.stream.Stream;

import javax.imageio.ImageIO;

import ru.cleverpumpkin.badge.filter.BadgeFilter;

public class ImageWriter {

    private final File outputFile;

    private final BufferedImage image;

    public ImageWriter(File inputFile, File outputFile) throws IOException {
        this.outputFile = outputFile;
        image = ImageIO.read(inputFile);
    }

    public void write() throws IOException {
        //noinspection ResultOfMethodCallIgnored
        outputFile.getParentFile().mkdirs();
        ImageIO.write(image, "png", outputFile);
    }

    public void process(Stream<BadgeFilter> filters) {
        filters.forEach(filter -> filter.apply(image));
    }
}

File: plugin/src/main/java/ru/cleverpumpkin/appbadge/utils/ImageWriter.kt
package ru.cleverpumpkin.appbadge.utils

import ru.cleverpumpkin.appbadge.filter.AppBadgeFilter
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import javax.imageio.ImageIO

class ImageWriter(private val outputFile: File) {

    private var image: BufferedImage? = null

    @Throws(IOException::class)
    fun read(inputFile: File) {
        image = ImageIO.read(inputFile)
    }

    @Throws(IOException::class)
    fun write() {
        outputFile.parentFile.mkdirs()
        ImageIO.write(image, "png", outputFile)
    }

    fun process(filters: List<AppBadgeFilter>, isAdaptive: Boolean) {
        if (image == null) {
            // start with empty transparent image
            image = BufferedImage(1024, 1024, BufferedImage.TYPE_INT_ARGB)
        }
        image?.let { image ->
            filters.forEach { filter -> filter.apply(image, isAdaptive) }
        }
    }
}